pkg/build/gather-build-info.go

```diff
@@ -0,0 +1,147 @@
+	AppBundleVersion  string            `json:"appBundleVersion,omitempty"`
+	SourceControl     SourceControl     `json:"sourceControl,omitempty"`
+	BuilderName       string            `json:"builderName,omitempty"`
+	ReleaseStage      string            `json:"releaseStage,omitempty"`
+	AppVersion        string            `json:"appVersion,omitempty"`
+	AutoAssignRelease *bool             `json:"autoAssignRelease,omitempty"`
+	MetaData          map[string]string `json:"metadata,omitempty"`
+}
+
+func (opts CreateBuildInfo) Override(base CreateBuildInfo) CreateBuildInfo {
+	return CreateBuildInfo{
+		ApiKey:           utils.ThisOrThat(opts.ApiKey, base.ApiKey).(string),
+		AppVersionCode:   utils.ThisOrThat(opts.AppVersionCode, base.AppVersionCode).(string),
+		AppBundleVersion: utils.ThisOrThat(opts.AppBundleVersion, base.AppBundleVersion).(string),
+		SourceControl: SourceControl{
+			Provider:   utils.ThisOrThat(opts.SourceControl.Provider, base.SourceControl.Provider).(string),
+			Repository: utils.ThisOrThat(opts.SourceControl.Repository, base.SourceControl.Repository).(string),
+			Revision:   utils.ThisOrThat(opts.SourceControl.Revision, base.SourceControl.Revision).(string),
+		},
+		BuilderName:       utils.ThisOrThat(opts.BuilderName, base.BuilderName).(string),
+		ReleaseStage:      utils.ThisOrThat(opts.ReleaseStage, base.ReleaseStage).(string),
+		AppVersion:        utils.ThisOrThat(opts.AppVersion, base.AppVersion).(string),
+		AutoAssignRelease: utils.ThisOrThatBool(opts.AutoAssignRelease, base.AutoAssignRelease),
+		MetaData:          utils.ThisOrThat(opts.MetaData, base.MetaData).(map[string]string),
+	}
+}
+
+func (opts CreateBuildInfo) Validate() error {
+	if opts.ApiKey == "" {
+		return fmt.Errorf("missing api key, please specify using `--api-key`")
+	}
+
+	if opts.AppVersion == "" {
+		return fmt.Errorf("missing app version, please specify using `--app-version`")
+	}
+
+	return nil
+}
+
+func PopulateFromCliOpts(opts options.CLI) CreateBuildInfo {
+	return CreateBuildInfo{
+		ApiKey:           opts.ApiKey,
+		AppVersionCode:   opts.CreateBuild.AndroidBuildOptions.VersionCode,
+		AppBundleVersion: opts.CreateBuild.IosBuildOptions.BundleVersion,
+		SourceControl: SourceControl{
+			Provider:   opts.CreateBuild.Provider,
+			Repository: opts.CreateBuild.Repository,
+			Revision:   opts.CreateBuild.Revision,
+		},
+		BuilderName:       opts.CreateBuild.BuilderName,
+		ReleaseStage:      opts.CreateBuild.ReleaseStage,
+		AppVersion:        opts.CreateBuild.VersionName,
+		AutoAssignRelease: &opts.CreateBuild.AutoAssignRelease,
+		MetaData:          opts.CreateBuild.Metadata,
+	}
+}
+
+func PopulateFromPath(path string) CreateBuildInfo {
+	return CreateBuildInfo{
+		ApiKey:           "",
+		AppVersionCode:   "",
+		AppBundleVersion: "",
+		SourceControl: SourceControl{
+			Provider:   "",
+			Repository: utils.GetRepoUrl(path),
+			Revision:   utils.GetCommitHash(),
```

Comment:
There's somewhat of an assumption that they are using git. Does this fail OK if that is not the case? ... or if git isn't on the path?

I think it would also be good to rename the functions to make it clear this is git only.